bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/internal/CommunicationManager.java

```diff
@@ -43,144 +58,223 @@
-                                        handler.getThing().getUID(), SafeMethodCaller.DEFAULT_TIMEOUT);
-                            } catch (Exception ex) {
-                                logger.error("Exception occurred while calling handler: {}", ex.getMessage(), ex);
-                            }
-                        } else {
-                            logger.debug("Not delegating command '{}' for item '{}' to handler for channel '{}', "
-                                    + "because handler is not initialized (thing must be in status UNKNOWN, ONLINE or OFFLINE).",
-                                    command, itemName, channelUID);
-                        }
-                    } else {
-                        logger.warn("Cannot delegate command '{}' for item '{}' to handler for channel '{}', "
-                                + "because no handler is assigned. Maybe the binding is not installed or not "
-                                + "propertly initialized.", command, itemName, channelUID);
-                    }
-                } else {
-                    logger.warn(
-                            "Cannot delegate command '{}' for item '{}' to handler for channel '{}', "
-                                    + "because no thing with the UID '{}' could be found.",
-                            command, itemName, channelUID, channelUID.getThingUID());
+    private Profile getProfile(ItemChannelLink link, Item item, Thing thing) {
+        if (thing == null) {
+            return new NoOpProfile();
+        }
+
+        Channel channel = thing.getChannel(link.getLinkedUID().getId());
+        if (channel == null) {
+            return new NoOpProfile();
+        }
+
+        Profile profile = null;
+        synchronized (profiles) {
+            profile = profiles.get(link.getLinkedUID());
+            if (profile == null) {
+                profile = getProfileFromFactories(link, item, channel);
+                if (profile == null) {
+                    logger.trace("No profile factory found for link {}, falling back to the defaults", link);
+                    profile = createDefaultProfile(channel);
+                }
+                if (profile != null) {
+                    profiles.put(link.getLinkedUID(), profile);
                 }
             }
         }
+        return profile != null ? profile : new NoOpProfile();
     }
 
-    @Override
-    protected void receiveUpdate(ItemStateEvent updateEvent) {
-        String itemName = updateEvent.getItemName();
+    private Profile getProfileFromFactories(ItemChannelLink link, Item item, Channel channel) {
+        for (ProfileFactory profileFactory : profileFactories.keySet()) {
+            Profile profile = profileFactory.createProfile(link, item, channel);
+            if (profile != null) {
+                profileFactories.get(profileFactory).add(link.getLinkedUID());
```

Comment:
If you need access to the profileFactory, wouldn't it make sense to use `entrySet()` instead of `keySet()`?